### PR TITLE
Expose CUDAPinnedPlace to Python

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/9658

### CODE REVIEW

File: paddle/fluid/framework/tensor_impl.h [hunk 1]
```diff
@@ -128,13 +128,21 @@ inline void* Tensor::mutable_data(platform::Place place, std::type_index type) {
     if (platform::is_cpu_place(place)) {
       holder_.reset(new PlaceholderImpl<platform::CPUPlace>(
           boost::get<platform::CPUPlace>(place), size, type));
-    } else if (platform::is_gpu_place(place)) {
+    } else if (platform::is_gpu_place(place) ||
+               platform::is_cuda_pinned_place(place)) {
 #ifndef PADDLE_WITH_CUDA
-      PADDLE_THROW("'CUDAPlace' is not supported in CPU only device.");
+      PADDLE_THROW(
+          "'CUDAPlace' or 'CUDAPinnedPlace' is not supported in CPU only "
```
Comment (wangkuiyi) [inline paddle/fluid/framework/tensor_impl.h]:
```c++
"CUDAPlace or CUDAPinnedPlace is not supported in CPU-only mode"
```

File: paddle/fluid/pybind/tensor_py.h [hunk 2]
```diff
@@ -208,6 +210,38 @@ void PyCUDATensorSetFromArray(
...
+  for (size_t i = 0; i < array.ndim(); ++i) {
+    dims.push_back(static_cast<int>(array.shape()[i]));
+  }
+
+  self.Resize(framework::make_ddim(dims));
+  auto *dst = self.mutable_data<T>(place);
+  std::memcpy(dst, array.data(), sizeof(T) * array.size());
+}
+
+template <>
```
Comment (wangkuiyi) [inline paddle/fluid/pybind/tensor_py.h]: I think we need a comment explaining the reason that we need this specialization:

```c++
// This following specialization maps uint16_t in the parameter type to platform::float16.

File: paddle/fluid/framework/tensor_impl.h (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/tensor_impl.h]: Done

File: paddle/fluid/pybind/tensor_py.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/pybind/tensor_py.h]: Thanks!  done

**Pull request merged**